### PR TITLE
Bug 1803844: Support private IPv6 address configurations

### PR DESCRIPTION
**Pull request by bison**
This adds support for detecting IPv6 address space on the subnet for a
new machine, and automatically adding a private IP configuration with
a dynamic address from that space.

### CODE REVIEW

File: pkg/cloud/azure/services/networkinterfaces/networkinterfaces.go [hunk 1]
```diff
@@ -234,3 +254,29 @@ func (s *Service) Delete(ctx context.Context, spec azure.Spec) error {
 	klog.V(2).Infof("successfully deleted nic %s", nicSpec.Name)
 	return err
 }
+
+func subnetHasIPv6(subnet network.Subnet) bool {
```
Comment (bison) [inline pkg/cloud/azure/services/networkinterfaces/networkinterfaces.go]: I don't know if this is the right thing, but I don't see any other way to tell if a subnet is IPv6-enabled.

Another option is to make the decision on whether to configure an IPv6 address explicit in the provider config. If the subnet doesn't have IPv6 address space, creating the NIC will fail and the Machine will go into a failed state.

**Comment by enxebre** [comment]
master tracks 4.5 now. We need a cherry-pick for 4.4 as well
/cherry-pick release-4.4

**Comment by danwinship** [comment]
Doesn't seem to have worked:

```
NAME                                             STATUS     ROLES    AGE   VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                                                       KERNEL-VERSION                CONTAINER-RUNTIME
dwinship-ipv6-43-xbn4q-master-0                  Ready      master   24m   v1.16.2   fc00::4       <none>        Red Hat Enterprise Linux CoreOS 43.81.202002170853.0 (Ootpa)   4.18.0-147.5.1.el8_1.x86_64   cri-o://1.16.3-22.dev.rhaos4.3.git11c04e3.el8
dwinship-ipv6-43-xbn4q-master-1                  Ready      master   24m   v1.16.2   fc00::6       <none>        Red Hat Enterprise Linux CoreOS 43.81.202002170853.0 (Ootpa)   4.18.0-147.5.1.el8_1.x86_64   cri-o://1.16.3-22.dev.rhaos4.3.git11c04e3.el8
dwinship-ipv6-43-xbn4q-master-2                  Ready      master   24m   v1.16.2   fc00::5       <none>        Red Hat Enterprise Linux CoreOS 43.81.202002170853.0 (Ootpa)   4.18.0-147.5.1.el8_1.x86_64   cri-o://1.16.3-22.dev.rhaos4.3.git11c04e3.el8
dwinship-ipv6-43-xbn4q-worker-centralus1-rvck7   NotReady   worker   14m   v1.16.2   10.0.32.6     <none>        Red Hat Enterprise Linux CoreOS 43.81.202002170853.0 (Ootpa)   4.18.0-147.5.1.el8_1.x86_64   cri-o://1.16.3-22.dev.rhaos4.3.git11c04e3.el8
dwinship-ipv6-43-xbn4q-worker-centralus2-m5k8j   NotReady   worker   14m   v1.16.2   10.0.32.5     <none>        Red Hat Enterprise Linux CoreOS 43.81.202002170853.0 (Ootpa)   4.18.0-147.5.1.el8_1.x86_64   cri-o://1.16.3-22.dev.rhaos4.3.git11c04e3.el8
dwinship-ipv6-43-xbn4q-worker-centralus3-6p246   NotReady   worker   14m   v1.16.2   10.0.32.4     <none>        Red Hat Enterprise Linux CoreOS 43.81.202002170853.0 (Ootpa)   4.18.0-147.5.1.el8_1.x86_64   cri-o://1.16.3-22.dev.rhaos4.3.git11c04e3.el8
```

**Comment by danwinship** [comment]
OK, ignore the previous comment; I hadn't overridden the image correctly. Now it does work:

```
NAME                                             STATUS   ROLES    AGE     VERSION   INTERNAL-IP     EXTERNAL-IP   OS-IMAGE                                                       KERNEL-VERSION                CONTAINER-RUNTIME
dwinship-ipv6-43-tvkrk-master-0                  Ready    master   16m     v1.16.2   fc00::7         <none>        Red Hat Enterprise Linux CoreOS 43.81.202002170853.0 (Ootpa)   4.18.0-147.5.1.el8_1.x86_64   cri-o://1.16.3-22.dev.rhaos4.3.git11c04e3.el8
dwinship-ipv6-43-tvkrk-master-1                  Ready    master   16m     v1.16.2   fc00::6         <none>        Red Hat Enterprise Linux CoreOS 43.81.202002170853.0 (Ootpa)   4.18.0-147.5.1.el8_1.x86_64   cri-o://1.16.3-22.dev.rhaos4.3.git11c04e3.el8
dwinship-ipv6-43-tvkrk-master-2                  Ready    master   16m     v1.16.2   fc00::4         <none>        Red Hat Enterprise Linux CoreOS 43.81.202002170853.0 (Ootpa)   4.18.0-147.5.1.el8_1.x86_64   cri-o://1.16.3-22.dev.rhaos4.3.git11c04e3.el8
dwinship-ipv6-43-tvkrk-worker-centralus1-42hbq   Ready    worker   6m9s    v1.16.2   fc00:0:0:1::6   <none>        Red Hat Enterprise Linux CoreOS 43.81.202002170853.0 (Ootpa)   4.18.0-147.5.1.el8_1.x86_64   cri-o://1.16.3-22.dev.rhaos4.3.git11c04e3.el8
dwinship-ipv6-43-tvkrk-worker-centralus2-mc4hb   Ready    worker   6m23s   v1.16.2   fc00:0:0:1::5   <none>        Red Hat Enterprise Linux CoreOS 43.81.202002170853.0 (Ootpa)   4.18.0-147.5.1.el8_1.x86_64   cri-o://1.16.3-22.dev.rhaos4.3.git11c04e3.el8
dwinship-ipv6-43-tvkrk-worker-centralus3-2hcmv   Ready    worker   6m58s   v1.16.2   fc00:0:0:1::4   <none>        Red Hat Enterprise Linux CoreOS 43.81.202002170853.0 (Ootpa)   4.18.0-147.5.1.el8_1.x86_64   cri-o://1.16.3-22.dev.rhaos4.3.git11c04e3.el8
```

The numbering of the workers doesn't quite match the numbering of the masters (`fc00:0:0:1::...` vs `fc00::...`) but it is still within the configured `machineNetwork` (`fc00::/48`). (Actually, the `/48` is kind of a mistake in my configuration; it really ought to be a `/64`.)

**Comment by danwinship** [comment]
> The numbering of the workers doesn't quite match the numbering of the masters (`fc00:0:0:1::...` vs `fc00::...`) but it is still within the configured `machineNetwork` (`fc00::/48`).

ok, and that's consistent with the IPv4 behavior too (where the workers get `10.0.32.n` rather than `10.0.0.n`). I guess it's just setting a random high bit to ensure that it doesn't reuse the same addresses the installer used...

**Comment by danwinship** [comment]
@enxebre @michaelgugino Brad had left this marked as "WIP" but the patch has been tested now and works so it would be good to get review before he gets back from PTO...

**Comment by danwinship** [comment]
/retest
(The lack of this fix is continuing to make IPv6-on-Azure development difficult since we have to maintain builds containing this patch.)

**Comment by bison** [comment]
@danwinship: The tests can't pass because Azure support is broken in installer master:

https://bugzilla.redhat.com/show_bug.cgi?id=1809346
https://github.com/openshift/installer/pull/3251

**Comment by enxebre** [comment]
can we ref bz https://bugzilla.redhat.com/show_bug.cgi?id=1803832

**Pull request merged**